### PR TITLE
[RFR] Complete lib

### PR DESCRIPTION
**Pull request by ThieryMichel**
follow #2

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    Generate challengeResponse from the key sent by the client during login. challengeResponse.challenge should be stocked serverside linked to the corresponding user
+- verifyAuthenticatorAssertion
+    Take the loginChallenge request body and the key stocked with the user, and return true if it passes the authenticator assertion
+
+See an example in example/server
+
+
+## Roadmap
+
+For now only fido-u2f and packed format are implemeted
```
Comment (djhi) [inline README.md]: s/implemeted/implemented/

File: README.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    parseRegisterRequest,
+    generateRegistrationChallenge,
+    parseLoginRequest,
+    generateLoginChallenge,
+    verifyAuthenticatorAssertion,
+} from '@webauthn/server';
+```
+
+- parseRegisterRequest:
+    Extract challenge and key from the register request body. The challenge allow to retrieve the user, and the key must be stocked server side linked to the user.
```
Comment (djhi) [inline README.md]: s/stocked/stored/

File: README.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+- generateRegistrationChallenge:
+    Generate a challenge from a relying party and an user `{ relyingParty, user }` to be sent back to the client, in order to register
```
Comment (djhi) [inline README.md]: s/an user/a user/

File: README.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+- parseLoginRequest:
+    Extract challenge and KeyId from the login request.
+- generateLoginChallenge:
+    Generate challengeResponse from the key sent by the client during login. challengeResponse.challenge should be stocked serverside linked to the corresponding user
```
Comment (djhi) [inline README.md]: s/stocked/stored/

File: README.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+- verifyAuthenticatorAssertion
+    Take the loginChallenge request body and the key stocked with the user, and return true if it passes the authenticator assertion
```
Comment (djhi) [inline README.md]: s/stocked/stored/

File: packages/server/src/authenticatorKey/parseFidoPackedKey.js [hunk 6]
```diff
@@ -95,23 +109,35 @@ exports.parseFidoPackedKey = (authenticatorKey, clientDataJSON) => {
...
+        'SHA256',
+        Buffer.from(clientDataJSON, 'base64')
+    );
+    const signatureBaseBuffer = Buffer.concat([
+        authenticatorDataBuffer,
+        clientDataHash,
+    ]);
 
-    const publicKey = convertASN1toPEM(Buffer.from(key.publicKey, 'base64'));
+    const publicKey = ConvertASN1toPEM(Buffer.from(key.publicKey, 'base64'));
```
Comment (djhi) [inline packages/server/src/authenticatorKey/parseFidoPackedKey.js]: Why pascal case ? Isn't it a function ?

File: packages/server/src/authenticatorKey/parseFidoPackedKey.js [hunk 7]
```diff
@@ -222,11 +248,22 @@ exports.validateFidoPackedKey = (authenticatorDataBuffer, key, clientDataJSON, b
...
-function verifyFullAttestation(authenticatorKey, signatureBaseBuffer, signatureBuffer) {
-    const leafCert = convertASN1toPEM(authenticatorKey.attStmt.x5c[0]);
+function verifyFullAttestation(
+    authenticatorKey,
+    signatureBaseBuffer,
+    signatureBuffer
+) {
+    const authenticatorData = parseAttestationData(authenticatorKey.authData);
+
+    const publicKey = ConvertCOSEPublicKeyToRawPKCSECDHAKey(
```
Comment (djhi) [inline packages/server/src/authenticatorKey/parseFidoPackedKey.js]: Same here about Pascal case

File: packages/server/src/authenticatorKey/parseFidoPackedKey.js [hunk 8]
```diff
@@ -222,11 +248,22 @@ exports.validateFidoPackedKey = (authenticatorDataBuffer, key, clientDataJSON, b
...
+        authenticatorData.COSEPublicKey
+    );
+    const leafCert = ConvertASN1toPEM(authenticatorKey.attStmt.x5c[0]);
```
Comment (djhi) [inline packages/server/src/authenticatorKey/parseFidoPackedKey.js]: here too

File: packages/server/src/authenticatorKey/parseFidoU2FKey.js [hunk 9]
```diff
@@ -8,18 +13,23 @@ exports.parseFidoU2FKey = (authenticatorKey, clientDataJSON) => {
...
     }
 
-    const clientDataHash = hash('SHA256', Buffer.from(clientDataJSON, 'base64'));
+    const clientDataHash = hash(
+        'SHA256',
+        Buffer.from(clientDataJSON, 'base64')
+    );
     const reservedByte = Buffer.from([0x00]);
-    const publicKey = convertCOSEPublicKeyToRawPKCSECDHAKey(authenticatorData.COSEPublicKey);
+    const publicKey = ConvertCOSEPublicKeyToRawPKCSECDHAKey(
```
Comment (djhi) [inline packages/server/src/authenticatorKey/parseFidoU2FKey.js]: pascal case

File: packages/server/src/authenticatorKey/parseFidoU2FKey.js [hunk 10]
```diff
@@ -8,18 +13,23 @@ exports.parseFidoU2FKey = (authenticatorKey, clientDataJSON) => {
...
         reservedByte,
         authenticatorData.rpIdHash,
         clientDataHash,
         authenticatorData.credID,
-        publicKey
+        publicKey,
     ]);
 
-    const PEMCertificate = convertASN1toPEM(authenticatorKey.attStmt.x5c[0]);
+    const PEMCertificate = ConvertASN1toPEM(authenticatorKey.attStmt.x5c[0]);
```
Comment (djhi) [inline packages/server/src/authenticatorKey/parseFidoU2FKey.js]: pascal case

File: packages/server/src/authenticatorKey/parseFidoU2FKey.js [hunk 11]
```diff
@@ -29,35 +39,43 @@ exports.parseFidoU2FKey = (authenticatorKey, clientDataJSON) => {
...
     const signatureBase = Buffer.concat([
         authenticatorData.rpIdHash,
         authenticatorData.flagsBuf,
         authenticatorData.counterBuf,
-        clientDataHash
+        clientDataHash,
     ]);
 
-    const publicKey = convertASN1toPEM(Buffer.from(key.publicKey, 'base64'));
+    const publicKey = ConvertASN1toPEM(Buffer.from(key.publicKey, 'base64'));
```
Comment (djhi) [inline packages/server/src/authenticatorKey/parseFidoU2FKey.js]: pascal case

File: packages/server/src/utils.js [hunk 12]
```diff
@@ -1,29 +1,24 @@
...
+    return createHash(alg)
         .update(data)
         .digest();
+};
 
 /**
  * Convert binary certificate or public key to an OpenSSL-compatible PEM text format.
  */
-exports.convertASN1toPEM = pkBuffer => {
+exports.ConvertASN1toPEM = pkBuffer => {
```
Comment (djhi) [inline packages/server/src/utils.js]: pascal case

File: packages/server/src/utils.js [hunk 13]
```diff
@@ -64,10 +62,45 @@ exports.convertASN1toPEM = pkBuffer => {
     return `-----BEGIN ${type}-----\n` + PEMKey + `\n-----END ${type}-----\n`;
 };
 
+/**
+ * Takes COSE encoded public key and converts it to RAW PKCS ECDHA key
+ * @param  {Buffer} cosePublicKey - COSE encoded public key
+ * @return {Buffer}               - RAW PKCS encoded public key
+ */
+exports.ConvertCOSEPublicKeyToRawPKCSECDHAKey = cosePublicKey => {
```
Comment (djhi) [inline packages/server/src/utils.js]: pascal case

**Review by djhi — CHANGES_REQUESTED** [review]
Nice refactoring! Can you use ticks for names in the README ?

File: README.md (hunk 1 above)
Comment (ThieryMichel) [reply README.md]:
```suggestion
For now only fido-u2f and packed format are implemented
```

File: README.md (hunk 2 above)
Comment (ThieryMichel) [reply README.md]:
```suggestion
    Extract challenge and key from the register request body. The challenge allow to retrieve the user, and the key must be stored server side linked to the user.
```

File: README.md (hunk 3 above)
Comment (ThieryMichel) [reply README.md]:
```suggestion
    Generate a challenge from a relying party and a user `{ relyingParty, user }` to be sent back to the client, in order to register
```

File: README.md (hunk 5 above)
Comment (ThieryMichel) [reply README.md]:
```suggestion
    Take the loginChallenge request body and the key stored with the user, and return true if it passes the authenticator assertion
```

File: README.md (hunk 4 above)
Comment (ThieryMichel) [reply README.md]:
```suggestion
    Generate challengeResponse from the key sent by the client during login. challengeResponse.challenge should be stored serverside linked to the corresponding user
```

File: packages/server/src/authenticatorKey/parseFidoPackedKey.js (hunk 6 above)
Comment (ThieryMichel) [reply packages/server/src/authenticatorKey/parseFidoPackedKey.js]: It's acronym ASN1 (Abstract Syntax Notation One) and PEM (Privacy Enhanced Mail)

Comment (ThieryMichel) [reply packages/server/src/authenticatorKey/parseFidoPackedKey.js]: oups did not see the uppercase C

**Review by djhi — APPROVED** [review]

**Pull request merged**